### PR TITLE
Blog Content: Demystifying Istio's Sidecar Injection Model

### PR DESCRIPTION
**Pull request by chugtum**
This is a technical blog post, which helps de-mystify how Istio manages to plugin its data-plane components into an existing k8s deployment. Basically, how the sidecar injection is done whether manually or namespaceSelector based automatic injection. And also how the iptables are modified internally in the pod namespace to redirect the traffic to the injected proxy. This is 'behind the scenes` technical content and I believe it will help users in general who are trying to either understand about the concept or trying out Istio, first time.

I think there are many interested users who are looking for a structured deep-dive into Istio 'behind-the-scenes' at a single place and this is the starting post to accomplish the same.

### CODE REVIEW

**Comment by chugtum** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).



> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).

`I signed it`. It is the company's corporate CLA, with the e-mail group:
opensrc-contributors@avinetworks.com

**Comment by chugtum** [comment]
The CircleCI failure seems to be an already existing issue on the main branch.
./public/about/contribute/style-guide/index.html
External link http://stephanieleary.com/2015/05/why-click-here-is-a-terrible-link-and-what-to-write-instead/ failed: 500 No error

**Comment by rcaballeromx** [comment]
@chugtum Do you want me to perform a detailed review? 
I took a quick look and would advise you:
* Mind the 78 character line length.
* Some of your code blocks are very long. Consider just including the relevant parts or having things like the templates on separate files and link to them.

File: content/blog/2018/data-plane-setup/index.md [hunk 1]
```diff
@@ -0,0 +1,533 @@
+---
+title: Demystifying Istio's Sidecar Injection Model 
+description: De-mystify how Istio manages to plugin its data-plane components into an existing deployment. Deep-dive into sidecar injection, how it is done whether manually or `namespaceSelector` based automatic injection. And also how the iptables are modified in the pod namespace to redirect the traffic to the injected proxy.
```
Comment (geeknoid) [inline content/blog/2018/data-plane-setup/index.md]: Please keep descriptions to a single sentence. This should be plain text, no markdown.

File: content/blog/2018/data-plane-setup/index.md [hunk 2]
```diff
@@ -0,0 +1,533 @@
+---
+title: Demystifying Istio's Sidecar Injection Model 
+description: De-mystify how Istio manages to plugin its data-plane components into an existing deployment. Deep-dive into sidecar injection, how it is done whether manually or `namespaceSelector` based automatic injection. And also how the iptables are modified in the pod namespace to redirect the traffic to the injected proxy. 
+publishdate: 2018-12-12
+subtitle:
+attribution: Manish CHUGTU
+twitter: chugtum 
+weight: 78
```
Comment (geeknoid) [inline content/blog/2018/data-plane-setup/index.md]: This number will need to be adjusted before publishing to ensure proper ordering on the site.

File: content/blog/2018/data-plane-setup/index.md [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+
+{{< image width="40%"
+    ratio="33%"
+    link="./arch-2.svg"
+    alt="The overall architecture of an Istio-based application."
+    caption="Istio Architecture"
+    >}}
+It is important to understand that the sidecar injection into application pods happens automatically, though manual injection is also possible. Traffic is directed from the application services to and from these sidecars without developers needing to worry about it. Once they are connected to the Istio service mesh, they can start using and reaping the benefits of all that it has to offer. But how does the data plane plumbing happen and what is really required to make it work seamlessly? In this post we will deep-dive into the specifics of sidecar injection models for a very clear understanding of how it works.
+
+## Sidecar Injection
```
Comment (geeknoid) [inline content/blog/2018/data-plane-setup/index.md]: Here and elsewhere, only capitalize the first letter of headers (unless its a proper noun or acronym)

File: content/blog/2018/data-plane-setup/index.md [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+
+`istio-proxy`
+This is the actual sidecar proxy (based on Envoy).
+
+### Manual Injection
+
+In the manual injection method, `istioctl` can be used to modify the pod template and update it with the configuration of the above two containers. For both manual as well as automatic injection, Istio takes the configuration from the `istio-sidecar-injector` configmap and the mesh configmap `istio`.
+Let’s look at the configuration of `istio-sidecar-injector` configmap, to get an idea of what actually is going on.
+
+{{< text bash >}}
```
Comment (geeknoid) [inline content/blog/2018/data-plane-setup/index.md]: Make that "text bash yaml", which will cause the YAML output to be colorized.

File: content/blog/2018/data-plane-setup/index.md [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+  - name: istio-init
+    image: "gcr.io/istio-release/proxy_init:1.0.2"
+    args:
+    - "-p"
+    - [[ .MeshConfig.ProxyListenPort ]]
+{{< /text >}}
+
+You can also use the annotation `sidecar.istio.io/inject` in the pod template to override the default policy. The following is to disable automatic injection of the sidecar.
+
+{{< text plain>}}
```
Comment (geeknoid) [inline content/blog/2018/data-plane-setup/index.md]: plain -> yaml

File: static/js/all.min.js.map [hunk 6]
```diff
@@ -1 +1 @@
...
...AAS,eACTJ,YAAY,KAEhBrM,MAAME,UAAUf,aAAa,OAAQ,aAAc,CACnD0R,SAAU,CACNpE,QAAS,+CACTvD,MAAO,WACPiD,OAAQ,CAACkD,QAASrP,MAAME,UAAUyQ,KAAKtB,QAASD,YAAa,gBAGrEpP,MAAME,UAAU4Q,KAAO,CACnBC,SAAU,iCACV7B,OAAQ,CAACzC,QAAS,gCAAiCH,QAAQ,GAC3DkD,OAAQ,6DACRJ,YAAa,aACbK,SAAU,KACVH,QAAW,sBACX0B,KAAQ,aACThR,MAAME,UAAU+Q,MAAQjR,MAAME,UAAU4Q,KAC3C9Q,MAAME,UAAUgR,SAAWlR,MAAME,UAAU8J,OAAO,QAAS,CACvDqF,QAAS,sCACTqB,QAAS,8CACTS,UAAW,CAAC1E,QAAS,mHAAoHvD,MAAO,YAEpJlJ,MAAME,UAAUkR,OAAS,CACrBnC,QAAS,CAACxC,QAAS,eAAgBJ,YAAY,GAC/CgF,uBAAwB,CAAC5E,QAAS,sBAAuBH,QAAQ,EAAIpD,MAAO,UAC5EgG,OAAQ,CAACzC,QAAS,kCAAmCH,QAAQ,GAC7DiD,SAAY,CAAC9C,QAAS,4CAA6CJ,YAAY,GAC/E8C,aAAc,CAAC1C,QAAS,mBAAoBJ,YAAY,GACxDgD,QAAS,wLACTqB,QAAS,shBACTpB,QAAW,0BACXE,OAAQ,6FACRC,SAAU,uEACVL,YAAa,iBAEjBpP,MAAME,UAAUoR,KAAO,CACnBC,OAAQ,CACJ9E,QAAS,wFACTJ,YAAY,EACZnD,MAAO,UAEX+F,QAAS,MACTuC,IAAK,CAAC/E,QAAS,+EAAgFJ,YAAY,EAAInD,MAAO,UACtHuI,UAAW,CAAChF,QAAS,gBAAiBJ,YAAY,EAAInD,MAAO,aAC7DwI,SAAU,CACNjF,QAAS,2MACTJ,YAAY,EACZnD,MAAO,UAEXoG,QAAW,CAAC7C,QAAS,mEAAoEJ,YAAY,EAAInD,MAAO,aAChH8H,KAAQ,CAACvE,QAAS,+DAAgEJ,YAAY,EAAInD,MAAO,aACzGgG,OAAQ,CAACzC,QAAS,uFAAwFJ,YAAY,EAAIC,QAAQ,GAClIkD,OAAQ,CACJ/C,QAAS,gIACTJ,YAAY,GAEhB8B,IAAK,UACLwD,UAAW,YACXvC,YAAa"}
```
Comment (geeknoid) [inline static/js/all.min.js.map]: This file doesn't belong in this PR, please remove.

File: content/blog/2018/data-plane-setup/index.md (hunk 1 above)
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Done.

File: content/blog/2018/data-plane-setup/index.md (hunk 2 above)
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Got that - I am leaving it like that as of now.

File: content/blog/2018/data-plane-setup/index.md (hunk 3 above)
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Done at all the places.

File: content/blog/2018/data-plane-setup/index.md (hunk 4 above)
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Done.

File: content/blog/2018/data-plane-setup/index.md (hunk 5 above)
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Done.

File: static/js/all.min.js.map (hunk 6 above)
Comment (chugtum) [reply static/js/all.min.js.map]: I am not sure, how it got added. Is there an easy way to removing it from the PR. One way I can think of is, git checkout the one from istio.io/master and re-commit/push. Any suggestions here ?

**Comment by chugtum** [comment]
> @chugtum Do you want me to perform a detailed review?
> I took a quick look and would advise you:
> 
> * Mind the 78 character line length.
> * Some of your code blocks are very long. Consider just including the relevant parts or having things like the templates on separate files and link to them.

Thanks. Have taken care of the code block comment.

File: static/js/all.min.js.map (hunk 6 above)
Comment (chugtum) [reply static/js/all.min.js.map]: Done and we shouldn't be seeing this file in the diff now.

**Comment by geeknoid** [comment]
@rcaballeromx Can you take a full editorial pass?
@ayj Can you do a technical check?

**Comment by ayj** [comment]
cc @costinm for networking / iptables 
cc @ostromart for manual/automatic injection

**Comment by chugtum** [comment]
Hi Folks

Looks like, we are still pending some reviews on this. Was wondering, if you folks can spend some time to do the final set of reviews. Being a tech blog, there is a high chance that the content will get stale, if we can't publish it soon enough.

Thnx
Manish

**Comment by rcaballeromx** [comment]
I'm taking a full editorial pass on this as requested.
/hold

**Comment by chugtum** [comment]
Thanks Rigs.

**Comment by chugtum** [comment]
> I've completed the comprehensive review and left suggestions to address the issues I found.
> Most issues are about the clarity and understand-ability of the text. There are a few suggestions related to consistency too.

Hi  Rigs
Thnx for the comprehensive review. 
I have addressed all the changes as requested by you. Hope it looks fine now and ready to merge/publish.
 
-Manish

File: content/blog/2018/data-plane-setup/index.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
+---
+title: Demystifying Istio's Sidecar Injection Model 
+description: De-mystify how Istio manages to plugin its data-plane components into an existing deployment.
+publishdate: 2018-12-12
+subtitle:
+attribution: Manish CHUGTU
+twitter: chugtum 
+weight: 78
+keywords: [kubernetes, istio, sidecar-injection, admission-controller, mutatingwebhook]
```
[the remark this answers is not in the document]
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Done.

File: content/blog/2018/data-plane-setup/index.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+---
+A simple overview of an Istio service-mesh architecture always starts with describing the control-plane and data-plane.
+
+_From Istio’s Documentation:_
```
[the remark this answers is not in the document]
Comment (chugtum) [reply content/blog/2018/data-plane-setup/index.md]: Done

File: content/blog/2018/data-plane-setup/index.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
+-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
+-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
+-A ISTIO_OUTPUT -j ISTIO_REDIRECT
+-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
+{{< /text >}}
+
+From the output above, it is now clear that all the incoming traffic to port 80, which is where our application is listening is now `REDIRECTED` to port `15001`, which is the istio-proxy (envoy) listen port. The same holds true for the outgoing traffic too.
+
+This brings us to the end of this post. I hope it helped to de-mystify how Istio manages to plugin its data-plane components into an existing deployment, mainly injection of the sidecar containers and traffic routing to the proxy.
```
Comment (chugtum) [inline content/blog/2018/data-plane-setup/index.md]: Hi Rigs

Was wondering, if you had a look at this and if it looks good to be merged.

Thnx
Manish

**Comment by rcaballeromx** [comment]
@geeknoid Is there anything we can do to address the remaining unsuccessful checks?
@chugtum I'll try to take a look at your updates today.

**Comment by geeknoid** [comment]
To fix the CI failures, please rebase and repush the PR.

Thanks.

**Comment by chugtum** [comment]
> To fix the CI failures, please rebase and repush the PR.
> 
> Thanks.

@geeknoid  @rcaballeromx 
Thanks. Have re-based and re-pushed and CI is passing now. 

For CLA - Not sure why that is failing. I have signed it (initially as corporate and have that e-mail added to my github account too and since it was still throwing failure, have done it using individual too). Can you kindly confirm ?

Thnx

**Comment by chugtum** [comment]
> @geeknoid Is there anything we can do to address the remaining unsuccessful checks?
> @chugtum I'll try to take a look at your updates today.

Thanks @rcaballeromx . I have re-based and that fixed the CI check. For the cla check, can you kindly help as I have signed the agreement both as corporate and individual and used those e-mails in my git commit. So not sure, if you can manually verify and help get this PR merged.

**Comment by geeknoid** [comment]
As soon as Rigs is happy with the PR, I'll force-merge it. Sometimes the
CLA stuff gets stuck and it's easier to just ignore it :-)

**Comment by chugtum** [comment]
Thanks Martin - Appreciate that.

File: content/blog/2018/data-plane-setup/index.md [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+To use modified configmaps or local configmaps:
+
+- Create `inject-config.yaml` and `mesh-config.yaml` from the configmaps
+
+    {{< text bash >}}
+$ kubectl -n istio-system get configmap istio-sidecar-injector -o=jsonpath='{.data.config}' > inject-config.yaml
+$ kubectl -n istio-system get configmap istio -o=jsonpath='{.data.mesh}' > mesh-config.yaml
+    {{< /text >}}
+
+- Modify the existing pod template (Example: demo-red.yaml in my case)
```
Comment (rcaballeromx) [inline content/blog/2018/data-plane-setup/index.md]: You don't need the parenthesis.
```suggestion
- Modify the existing pod template, in my case, `demo-red.yaml`:
```

File: content/blog/2018/data-plane-setup/index.md [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+
+    {{< text bash >}}
+$ istioctl kube-inject --injectConfigFile inject-config.yaml --meshConfigFile mesh-config.yaml --filename demo-red.yaml --output demo-red-injected.yaml
+    {{< /text >}}
+
+- Apply the demo-red-injected.yaml
```
Comment (rcaballeromx) [inline content/blog/2018/data-plane-setup/index.md]: Use back-ticks for filenames for emphasis and consistency:
```suggestion
- Apply the `demo-red-injected.yaml`
```

**Comment by chugtum** [comment]
> Only minor issues left. Great job and thank you for being patient.

Hi @rcaballeromx 

No issues at all. I am actually impressed by the detail of the review. 
Hope it looks perfect now and you are happy with it. In that case, we can ask @geeknoid to merge it, as he is awaiting your green signal. 

Thnx
Manish

**Comment by chugtum** [comment]
> Thank you @chugtum It looks good to me, only the number for the correct order needs to be updated.

Thanks @rcaballeromx. 

@geeknoid I am assuming that the number would be updated to the correct value by you, before merging.

**Comment by geeknoid** [comment]
One last change before this goes in. Please move the whole directory to the 2019 directory and update the publication date.

Oh, and you can set the weight value in the front-matter to 76.

Thanks.

**Comment by chugtum** [comment]
> One last change before this goes in. Please move the whole directory to the 2019 directory and update the publication date.
> 
> Oh, and you can set the weight value in the front-matter to 76.
> 
> Thanks.

Done. Have moved it to 2019 and done the required updates for date and weight. Hope it looks good to merge now :).

**Comment by chugtum** [comment]
Thanks Martin - Appreciate that.

**Pull request merged**